### DISCUSSION TITLE
[Latex] Changing the font size in code blocks

### DISCUSSION THREAD

**Discussion by michaelosthege** [executablebooks/meta#730] [Q&A]
Hello, I'm trying to change (reduce) the font size of my code blocks.

For example with this block:

````
```{code-block} python
---
name: code_helloworld
linenos: True
caption: |
    This is just an example.
---
print("hello world")
````

Here's what I tried so far:
* Adding `fontsize: 6`, `font-size: 6`, or `fontSize: 6` to the directive parameters did not help.
* I tried to include the snippet from https://de.overleaf.com/learn/latex/Code_listing#Code_styles_and_colours in my preamble, but I couldn't get it to build.
* I tried to put the following before the code block, but it didn't help either:

````
% Creates a custom role for inserting raw latex
```{role} raw-latex(raw)
:format: latex
```

{raw-latex}`\newcommand{\listingsfont}{basicstyle=\ttfamily\tiny}`
````

Any ideas?

**Comment by mmcky** [comment]
This is a nice idea and it would be good to add font-size control. I am wondering though if this should be a global option that is set in `_config.yml` such that ALL code-cell font sizes are adjusted or if control at the code-cell level is preferred.

cc: @AakashGfude

**Reply by michaelosthege** [reply]
This should be a Sphinx setting, that I would assume exists already, but I couldn't find it.
Global would be nice. Local would be nice too.

I'll be happy as soon as either works

**Reply by mmcky** [reply]
I know there is sphinx font size setting for 10, 11 or 12pt fonts which can be set in the `_config.yml` using `latex_elements`

https://www.sphinx-doc.org/en/master/latex.html#the-latex-elements-configuration-setting

but that only gives you a narrow range of control. 

May need to add a new package the preamble to get other font sizes to work (more generally)

https://stackoverflow.com/questions/11958934/how-do-i-change-font-size-in-sphinx-latexpdf-with-anysize-or-cm-super

**Reply by michaelosthege** [reply]
That setting will effect everything, which doesn't help here.

I looked at the generated LaTeX and now I'm trying to override the `\sphinxVerbatim` command with a new font size..

Something like this.. It's __not working yet__ but I'd appreciate any help :)
I'll update this post if I can get it right.
```
\renewcommand{\sphinxVerbatim[1][1]}{\tiny{\OriginalsphinxVerbatim[#1]}}
```

**Reply by mmcky** [reply]
cc @AakashGfude any ideas on altering the `tex` directly to change the fonts in the code cells?

**Reply by AakashGfude** [reply]
@mmcky I think the most straightforward way would be to change font size of sphinxverbatim like @michaelosthege pointed out. As it is a wrapper around Verbatim from fancyvrb package, maybe we can try to find something from that package. I will have a look later.

**Reply by schmoelder** [reply]
Hey, I'm in a similar situation, is there any update on this issue?

**Comment by choldgraf** [comment]
For HTML, a straightforward approach would be to add a custom CSS file (https://jupyterbook.org/en/stable/advanced/html.html?highlight=css#custom-css-or-javascript)

With something like:

```css
div.highlight {
    font-size: 0.8em;
}
```

EDIT: just realized you're talking about Latex, I will update the title to clarify

**Comment by michaelosthege** [comment]
It took me about a day of trial and error, but I got it working.

Here's the part of my `_config.yml` that achieves the following three things:
1. Configure the text fontsize to `12pt` [as explained here](https://github.com/executablebooks/jupyter-book/issues/1640#issuecomment-1291237761)
2. Configure the line spacing to 1.5x
3. Set the font size used by `sphinxVerbatim`/`Verbatim` environments of code blocks.

Ultimately I found the solution by searching the [`fancyvrb.sty` source code](https://ctan.uib.no/usergrps/uktug/baskervi/7_1/fancyvrb.sty) for anything fontsize related, and then guessing how to override the command definition.

```yaml
sphinx:
  config:
    latex_elements: 
      papersize: "a4paper"
      pointsize: "12pt"  # Options: 10pt, 11pt, 12pt ONLY ⚠
      preamble: |
        \makeatletter % enable @-symbol for macros; don't forget to turn of with \makeatother

        % Use 1.5x lìne spacing
        \usepackage{setspace}
        \onehalfspacing

        % Reduce font ^size in ^code blocks
        % Override the fontsize command defined and used by fancyvrb (which implements Verbatim environments).
        % For size options see https://en.wikibooks.org/wiki/LaTeX/Fonts#Built-in_sizes
        \renewcommand{\FV@FontSize}{\scriptsize}

        \makeatother % disable @-symbol for macros
```

It's easy once you know how to do it 🤐
